File: Sorting/Medium/Radix Sort/radixsort.py
'''Copyrights to venkys.io
For more information, visite https://venkys.io"/

Python program to sort a given array in order using radix sort algorithm.'''

# Stable : Yes
# Inplace : Yes
# Adaptive : No

# Space complexity: O(n+k) where
  # n = number of elements in the input array
  # k = range of input values (number of unique digits)

#Time Complexity: O(d*(n+k)) /O(n+k) in simplified version
  # n=no. of elements
  # k= range of input values
  # d= no. of digits in maximum value

"""
    Perform counting sort on the array based on the given exponent.

    Parameters:
    arr (list): The input list to be sorted.
    exp (int): The exponent used for counting sort.

    Returns:
    None. Modifies the input array in-place.
    """
def counting_sort(arr, exp):
    n = len(arr)
    # Initialize an output array with zeros
    output = [0] * n
    # Initialize a count array to store the count of occurrences for each digit (0 to 9)
    count = [0] * 10

    # Count occurrences of each digit based on the given exponent
    for i in range(n):
        index = arr[i] // exp
        count[index % 10] += 1

    # Modify the count array to store the cumulative count of occurrences
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Build the output array using the count information
    i = n - 1
    while i >= 0:
        index = arr[i] // exp
        output[count[index % 10] - 1] = arr[i]
        count[index % 10] -= 1
        i -= 1

    # Copy the sorted elements from the output array to the original array
    for i in range(n):
        arr[i] = output[i]

"""
    Perform radix sort on the given array.

    Parameters:
    arr (list): The input list to be sorted.

    Returns:
    list: The sorted array.
    """
def radix_sort(arr):
    # Find the maximum number in the array to determine the number of digits in it
    max_num = max(arr)
    # Initialize the exponent (exp) to the least significant digit
    exp = 1
    # Loop until all digits are processed, from the least significant to the most significant
    while max_num // exp > 0:
        # Call the counting_sort function to sort the array based on the current exponent
        counting_sort(arr, exp)
        # Move to the next significant digit by multiplying the exponent by 10
        exp *= 10
    # Return the sorted array
    return arr


def main():
    try:
        arr = list(map(int, input("Enter the elements of the array separated by space: ").split()))
        # Check if the array is empty, and raise a ValueError if it is
        if len(arr) == 0:
            raise ValueError("Array is empty.")

        print('The given array is', arr)# Display the original array
        sorted_arr = radix_sort(arr.copy())  # Sorting a copy to keep the original array intact
        print('The sorted order is:', sorted_arr) # Display the sorted array

    except ValueError as e:
         # Handle ValueError exceptions (e.g., empty array)
        print(f"Error: {e}")

# Check if the script is being run as the main program
if __name__ == "__main__":
    main()

File: Sorting/Medium/Radix Sort/RadixSort.java
/* Copyrights to venkys.io */
/* For more programs, visit venkys.io */
/* Java program for Radix sort */

// Stable: Yes
// Inplace: No
// Adaptive: No

// Space complexity: O(n) where n = number of elements in the input array

// Time Complexity: O(d*(n+k)) / O(n+k) in simplified version
// n=no. of elements
// k= range of input values
// d= no. of digits in maximum value

import java.util.InputMismatchException;// Importing InputMismatchException class for handling input errors
import java.util.Scanner; // Importing Scanner class from java.util package for user input

public class RadixSort {

    static int output[]; // Sorted output array

    // Function to sort radixes
    public static void VSDRadix(int n, int[] a, int digits) {
        int k = 0;// Initialize the digit position to the least significant digit
        int temp[] = new int[n]; // Temporary array to hold radixes
        while (k < digits) { // Loop until all digits are processed
            for (int i = 0; i < n; i++) {
                if (k > 0)
                    // Extract the k-th digit using modulo and division
                    temp[i] = (int) ((a[i] / Math.pow(10, k)) % 10);
                else
                    // For the least significant digit, use modulo only
                    temp[i] = a[i] % 10;
            }
            // Perform counting sort based on the current digit
            VSDcountingsort(n, a, temp);
            k++; // Move to the next digit
        }
    }

    // Function to perform counting sort on radixes
    public static void VSDcountingsort(int n, int[] a, int temp[]) {
        int[] count = new int[10];// Create an array to store the count of occurrences of each radix.
        output = new int[n];// Create an array to store the sorted output.

        // Initializing all the elements of counting array to zero
        for (int i = 0; i < count.length; i++)
            count[i] = 0;

        // Storing the count of occurrences of each element from the input array to count array
        for (int i = 0; i < n; i++) {
            count[temp[i]]++;
        }

        // Computing cumulative sum from count array
        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        // Sorting radixes. Traverse the input array backward. 
        //For each element, use the count array to determine its position in the sorted output array.
        for (int i = n - 1; i >= 0; i--) {
            output[count[temp[i]] - 1] = a[i];
            count[temp[i]]--;//Decrease the count of that radix in the count array.
        }

        // Copy the sorted output array back to the original array.
        for (int i = 0; i < n; i++) {
            a[i] = output[i];
        }
    }

    // Function to count maximum number of digits
    public static int VSDdigitcount(int max) {
        int d = 0;// Initialize the digit count to 0
        while (max > 0) {
            d++; // Increment the digit count for each digit
            max = max / 10; // Remove the last digit by dividing the number by 10
        }
        return d; // Return the final count of digits
    }

    // Function to get maximum element in the array
    public static int VSDMax(int n, int[] a) {
        int max = 0;
        // Iterate through each element in the array
        for (int i = 0; i < n; i++) {
            // Check if the current element is greater than the current max
            if (a[i] > max) {
                // If yes, update max to the current element
                max = a[i];
            }
        }
        // Return the maximum element found in the array
        return max;
    }

    // Function to print the output
    public static void VSDprintOutput(int n, int[] a) {
        System.out.println("The sorted array is:");
        for (int i = 0; i < n; i++)
            System.out.print(a[i] + " ");

        System.out.println();
    }

    public static void main(String args[]) {
        // Create a Scanner object to read user input from the console.
        Scanner scanner = new Scanner(System.in);
        try {
            // Prompt the user to enter the size of the array.
            System.out.print("Enter the size of the array: ");
            // Initialize n to null to validate user input.
            Integer n = null;
             // Continue prompting the user until a valid positive integer is entered for the array size.
            while (n == null || n <= 0) {
                try {
                    // Attempt to read an integer from the user input.
                    n = scanner.nextInt();
                    // Check if the entered size is not positive.
                    if (n <= 0) {
                        System.out.println("Invalid size. Please enter a positive integer.");
                    }
                } catch (InputMismatchException e) {
                    // Handle the case where a non-integer value is entered.
                    System.out.println("Invalid input. Please enter a valid integer.");
                    scanner.nextLine();// Clear the buffer to prevent an infinite loop.
                }
            }

            // Create an array to store the elements of the array.
            int[] a = new int[n];
            
            // Prompt the user to enter the elements of the array.
            System.out.println("Enter the elements of the array:");
            // Iterate through each element of the array and input values from the user.
            for (int i = 0; i < n; i++) {
                // Use a loop to handle invalid input until a valid integer is entered.
                while (true) {
                    if (scanner.hasNextInt()) {
                        // If the entered value is an integer, assign it to the array.
                        a[i] = scanner.nextInt();
                        break; // Exit the loop as a valid integer is obtained.
                    } else {
                        // Display an error message for invalid input and clear the buffer.
                        System.out.println("Invalid input. Please enter valid integers.");
                        scanner.nextLine();// Clear the buffer to avoid an infinite loop.
                    }
                }
            }
            // Find the maximum element in the array
            int max = VSDMax(n, a);
            
            // Count the number of digits in the maximum element.
            int d = VSDdigitcount(max);
            
            // Perform Radix Sort on the array.
            VSDRadix(n, a, d);
            
            // Print the sorted array.
            VSDprintOutput(n, a);
        } finally {
            // Close the scanner to avoid resource leaks.
            if (scanner != null) {
                scanner.close();
            }
        }
    }
}
